File: javascript-calculator/calculator.js
class Calculator {
    constructor() {}

    getExpression(expressionString) {
        let ex = [];
        let value = "";

        for (let ch of expressionString) {
            if (ch === ' ') {
                continue;
            }
            if ('()^+-*/'.includes(ch)) {
                if (value !== "") {
                    let number = parseFloat(value);
                    if (isNaN(number)) {
                        throw new Error(`Could not parse ${value} into float`);
                    }
                    ex.push(number);
                    value = "";
                }
                ex.push(ch);
            } else {
                value += ch;
            }
        }

        if (value !== "") {
            let number = parseFloat(value);
            if (isNaN(number)) {
                throw new Error(`Could not parse ${value} into float`);
            }
            ex.push(number);
        }

        return ex;
    }

    getPrecedence(operation) {
        if (operation === '^') {
            return 3;
        } else if ('*/'.includes(operation)) {
            return 2;
        } else if ('+-'.includes(operation)) {
            return 1;
        } else {
            return 0;
        }
    }

    getAssociativity(operation) {
        if (operation === '^') {
            return 'R';
        }
        return 'L';
    }

    infixToPostfix(ex) {
        let result = [];
        let stack = [];

        for (let value of ex) {
            if (typeof value === 'number') {
                result.push(value);
            } else if (value === '(') {
                stack.push(value);
            } else if (value === ')') {
                while (stack.length > 0 && stack[stack.length - 1] !== '(') {
                    result.push(stack.pop());
                }
                if (stack.length === 0) {
                    throw new Error("Invalid expression, mismatched brackets");
                }
                stack.pop();
            } else {
                while (stack.length > 0 && stack[stack.length - 1] !== '(' &&
                    (this.getPrecedence(stack[stack.length - 1]) > this.getPrecedence(value) ||
                        (this.getPrecedence(stack[stack.length - 1]) === this.getPrecedence(value) &&
                            this.getAssociativity(value) === 'L'))) {
                    result.push(stack.pop());
                }
                stack.push(value);
            }
        }

        while (stack.length > 0) {
            if (stack[stack.length - 1] === '(') {
                throw new Error("Invalid expression, mismatched brackets");
            }
            result.push(stack.pop());
        }

        return result;
    }

    calculateArithmeticExpression(expressionString) {
        let start = process.hrtime.bigint();
        let ex = this.getExpression(expressionString);
        ex = this.infixToPostfix(ex);

        let stack = [];
        for (let value of ex) {
            if (value === '+') {
                if (stack.length < 2) {
                    throw new Error("Invalid expression: not enough operands for +");
                }
                let operand2 = stack.pop();
                let operand1 = stack.pop();
                stack.push(operand1 + operand2);
            } else if (value === '-') {
                if (stack.length < 2) {
                    throw new Error("Invalid expression: not enough operands for -");
                }
                let operand2 = stack.pop();
                let operand1 = stack.pop();
                stack.push(operand1 - operand2);
            } else if (value === '*') {
                if (stack.length < 2) {
                    throw new Error("Invalid expression: not enough operands for *");
                }
                let operand2 = stack.pop();
                let operand1 = stack.pop();
                stack.push(operand1 * operand2);
            } else if (value === '/') {
                if (stack.length < 2) {
                    throw new Error("Invalid expression: not enough operands for /");
                }
                let operand2 = stack.pop();
                let operand1 = stack.pop();
                stack.push(operand1 / operand2);
            } else if (value === '^') {
                if (stack.length < 2) {
                    throw new Error("Invalid expression: not enough operands for ^");
                }
                let operand2 = stack.pop();
                let operand1 = stack.pop();
                stack.push(Math.pow(operand1, operand2));
            } else {
                stack.push(value);
            }
        }

        if (stack.length !== 1) {
            throw new Error("Invalid expression: leftover operands");
        }

        let end = process.hrtime.bigint();
        console.log("Finished, took", (end - start).toString(), "nanoseconds");
        return stack[0];
    }
}

function main() {
    const calc = new Calculator();
    
    const readline = require('readline').createInterface({
        input: process.stdin,
        output: process.stdout
    });

    readline.setPrompt('arithmetic expression: ');

    readline.on('line', (expression) => {
        try {
            const result = calc.calculateArithmeticExpression(expression);
            console.log(`= ${result}`);
        } catch (error) {
            console.error("Error:", error.message);
        } finally {
            readline.prompt();
        }
    });

    readline.prompt();
}

if (require.main === module) {
    main();
}

File: python-calculator/calculator.py
import math
import time

class Calculator:
    def __init__(self):
        pass

    def get_expression(self, expression_string):
        ex = []
        value = ""

        for ch in expression_string:
            if ch == ' ':
                continue
            if ch in '()^+-*/':
                if value != "":
                    try:
                        number = float(value)
                    except ValueError:
                        raise ValueError(f"Could not parse {value} into float")
                    ex.append(number)
                    value = ""
                ex.append(ch)
            else:
                value += ch

        if value != "":
            try:
                number = float(value)
            except ValueError:
                raise ValueError(f"Could not parse {value} into float")
            ex.append(number)

        return ex

    def get_precedence(self, operation):
        if operation == '^':
            return 3
        elif operation in '*/':
            return 2
        elif operation in '+-':
            return 1
        else:
            return 0

    def get_associativity(self, operation):
        if operation == '^':
            return 'R'
        return 'L'

    def infix_to_postfix(self, ex):
        result = []
        stack = []

        for value in ex:
            if isinstance(value, float):
                result.append(value)
            elif value == '(':
                stack.append(value)
            elif value == ')':
                while stack and stack[-1] != '(':
                    result.append(stack.pop())
                if not stack:
                    raise ValueError("Invalid expression, mismatched brackets")
                stack.pop()
            else:
                while stack and stack[-1] != '(' and (
                    self.get_precedence(stack[-1]) > self.get_precedence(value) or 
                    (self.get_precedence(stack[-1]) == self.get_precedence(value) and self.get_associativity(value) == 'L')
                ):
                    result.append(stack.pop())
                stack.append(value)

        while stack:
            if stack[-1] == '(':
                raise ValueError("Invalid expression, mismatched brackets")
            result.append(stack.pop())

        return result

    def calculate_arithmetic_expression(self, expression_string):
        start = time.time_ns()
        ex = self.get_expression(expression_string)
        ex = self.infix_to_postfix(ex)

        stack = []
        for value in ex:
            if value == '+':
                if len(stack) < 2:
                    raise ValueError("Invalid expression: not enough operands for +")
                operand2 = stack.pop()
                operand1 = stack.pop()
                stack.append(operand1 + operand2)
            elif value == '-':
                if len(stack) < 2:
                    raise ValueError("Invalid expression: not enough operands for -")
                operand2 = stack.pop()
                operand1 = stack.pop()
                stack.append(operand1 - operand2)
            elif value == '*':
                if len(stack) < 2:
                    raise ValueError("Invalid expression: not enough operands for *")
                operand2 = stack.pop()
                operand1 = stack.pop()
                stack.append(operand1 * operand2)
            elif value == '/':
                if len(stack) < 2:
                    raise ValueError("Invalid expression: not enough operands for /")
                operand2 = stack.pop()
                operand1 = stack.pop()
                stack.append(operand1 / operand2)
            elif value == '^':
                if len(stack) < 2:
                    raise ValueError("Invalid expression: not enough operands for ^")
                operand2 = stack.pop()
                operand1 = stack.pop()
                stack.append(math.pow(operand1, operand2))
            else:
                stack.append(value)

        if len(stack) != 1:
            raise ValueError("Invalid expression: leftover operands")

        print("Finished, took", time.time_ns() - start)
        return stack[0]

def main():
    calc = Calculator()

    while True:
        try:
            expression = input("arithmetic expression: ")
            result = calc.calculate_arithmetic_expression(expression)
            print(f"= {result}")
        except KeyboardInterrupt:
            print("\nExiting.")
            break
        except Exception as e:
            print("Error:", e)

if __name__ == "__main__":
    main()

